src/plugins/discover/public/application/main/services/discover_shared_state_container.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import {
+  createStateContainer,
+  createStateContainerReactHelpers,
+  ReduxLikeStateContainer,
+} from '@kbn/kibana-utils-plugin/common';
+import { DataView, DataViewListItem } from '@kbn/data-views-plugin/common';
+import { DataTableRecord } from '../../../types';
+
+export interface DiscoverSharedState {
+  dataView: DataView | undefined;
+  savedDataViews: DataViewListItem[];
+  adHocDataViews: DataView[];
+  expandedDoc: DataTableRecord | undefined;
+}
+
+interface DiscoverSharedStateTransitions {
+  setDataView: (state: DiscoverSharedState) => (dataView: DataView) => DiscoverSharedState;
+  setSavedDataViews: (
+    state: DiscoverSharedState
+  ) => (dataView: DataViewListItem[]) => DiscoverSharedState;
+  setAdHocDataViews: (state: DiscoverSharedState) => (dataViews: DataView[]) => DiscoverSharedState;
+  appendAdHocDataViews: (
+    state: DiscoverSharedState
+  ) => (dataViews: DataView | DataView[]) => DiscoverSharedState;
+  removeAdHocDataViewById: (state: DiscoverSharedState) => (id: string) => DiscoverSharedState;
+  replaceAdHocDataViewWithId: (
+    state: DiscoverSharedState
+  ) => (id: string, dataView: DataView) => DiscoverSharedState;
+  setExpandedDoc: (
+    state: DiscoverSharedState
+  ) => (dataView: DataTableRecord | undefined) => DiscoverSharedState;
+}
+
+export type DiscoverSharedStateContainer = ReduxLikeStateContainer<
+  DiscoverSharedState,
+  DiscoverSharedStateTransitions
+>;
+
+export const {
+  Provider: DiscoverSharedStateProvider,
+  useSelector: useDiscoverSharedStateSelector,
+} = createStateContainerReactHelpers<ReduxLikeStateContainer<DiscoverSharedState>>();
+
+export function getSharedStateContainer() {
```

Comment:
Github seems to think this is a new file, but just the file and functions and types were renamed